helper/retry.go

```diff
@@ -0,0 +1,27 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package helper
+
+import (
+	"time"
+)
+
+// ExpiryToRenewTime calculates how long until clients should try to renew
+// credentials based on their expiration time and now.
+//
+// Renewals will begin halfway between now and the expiry plus some jitter.
+//
+// If the expiration is in the past or less than the min wait, then the min
+// wait time will be used with jitter.
```

Comment:
But doesn't this behavior floor the result at 15min? If so, when would the client ever renew if the time to renew is always 15min in the future? Is the assumption that clients will call this function (or something that calls it, like the JWKS endpoint) once and then set a timer on that value?

Reply:
> Is the assumption that clients will call this function (or something that calls it, like the JWKS endpoint) once and then set a timer on that value?

Yes, but oof you're right this function has a sort of Fence Post Problem: if you call it before each fetch to determine whether it's time to fetch: you'll never renew!

However if you call it _after you fetch_ it will give you a reasonable time to fetch in. JWKS is the latter, so I think it's safe:

1. Consul fetches JWKS v1 response, Cache-Control=15 minutes
2. Consul uses JWKS v1 for 15 minutes
3. Consul fetches JWKS v2 response, Cache-Control=...

If Consul was using something like ExpiryToRenewTime prior to each fetch I think we'd have the problem you mention.

I'll see if there's a safer way to write this function when I post the next PR that references it. It's possible we don't want to actually share this code at all and can just make it a private function for JWKS to use.